Enhancements to the openebs-provisioner

satyamz opened the pull request:
Why is this change required?
1. OpenEBS Provisioner should pass PVC namespace to maya_apiserver
2. Change VSM to Volume

What is changed?
1. VSM to Volume
2. OpenEBS Provisioner is passing PVC namespace to maya_apiserver

Fixes: 
- https://github.com/openebs/openebs/issues/784
- https://github.com/openebs/openebs/issues/787
- https://github.com/openebs/openebs/issues/727
- https://github.com/openebs/openebs/issues/810

Commit:
Enhancements to the openebs-provisioner

Why this change is required?
1. OpenEBS Provisioner should pass PVC namespace to maya_apiserver
2. Change VSM to Volume

What is changed?
1. VSM to Volume
2. OpenEBS Provisioner is passing PVC namespace to maya_apiserver

satyamz commented:
Output for Create Volume and Delete Volume:
```
ubuntu@kubemaster-01:~/openebs/k8s$ kubectl apply -f test-pvc.yaml 
persistentvolumeclaim "vut" created
ubuntu@kubemaster-01:~/openebs/k8s$ kubectl logs -f openebs-provisioner-3136189815-6v2h4
Error from server (NotFound): pods "openebs-provisioner-3136189815-6v2h4" not found
ubuntu@kubemaster-01:~/openebs/k8s$ k get pods 
NAME                                                             READY     STATUS    RESTARTS   AGE
maya-apiserver-1089964587-4c196                                  1/1       Running   0          1m
openebs-provisioner-3136189815-8wwnx                             1/1       Running   0          1m
pvc-9944178c-bd73-11e7-81db-021c6f7dbe9d-ctrl-3259361231-8dzh7   1/1       Running   0          12s
pvc-9944178c-bd73-11e7-81db-021c6f7dbe9d-rep-2256969920-mprr7    1/1       Running   0          11s
pvc-9944178c-bd73-11e7-81db-021c6f7dbe9d-rep-2256969920-v4580    1/1       Running   0          11s
ubuntu@kubemaster-01:~/openebs/k8s$ k logs -f openebs-provisioner-3136189815-8wwnx 
I1030 13:09:11.887559       1 maya_api.go:65] Maya Cluster IP: 10.111.244.51
I1030 13:09:11.915624       1 controller.go:407] Starting provisioner controller 85a0b3b4-bd73-11e7-b51f-469266a6fddd!
I1030 13:09:46.677218       1 controller.go:1073] scheduleOperation[lock-provision-default/vut[9944178c-bd73-11e7-81db-021c6f7dbe9d]]
I1030 13:09:46.681262       1 controller.go:1073] scheduleOperation[lock-provision-default/vut[9944178c-bd73-11e7-81db-021c6f7dbe9d]]
I1030 13:09:46.692432       1 leaderelection.go:156] attempting to acquire leader lease...
I1030 13:09:46.703453       1 leaderelection.go:178] successfully acquired lease to provision for pvc default/vut
I1030 13:09:46.703677       1 controller.go:1073] scheduleOperation[provision-default/vut[9944178c-bd73-11e7-81db-021c6f7dbe9d]]
I1030 13:09:46.712277       1 maya_api.go:87] Volume Spec Created:
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: pvc-9944178c-bd73-11e7-81db-021c6f7dbe9d
  labels:
    volumeprovisioner.mapi.openebs.io/storage-size: 5G
    k8s.io/storage-class: openebs-standard
    k8s.io/namespace: default

I1030 13:09:46.833081       1 maya_api.go:115] Volume Successfully Created:
{"metadata":{"creationTimestamp":null,"name":"pvc-9944178c-bd73-11e7-81db-021c6f7dbe9d"},"spec":{"AccessModes":null,"Capacity":null,"ClaimRef":null,"OpenEBS":{"volumeID":""},"PersistentVolumeReclaimPolicy":"","StorageClassName":""},"status":{"Message":"","Phase":"","Reason":""}}
I1030 13:09:46.833168       1 maya_api.go:130] Get details for Volume :pvc-9944178c-bd73-11e7-81db-021c6f7dbe9d
I1030 13:09:46.948478       1 maya_api.go:148] Volume Details Successfully Retrieved
I1030 13:09:46.948632       1 controller.go:806] volume "pvc-9944178c-bd73-11e7-81db-021c6f7dbe9d" for claim "default/vut" created
I1030 13:09:46.958759       1 controller.go:823] volume "pvc-9944178c-bd73-11e7-81db-021c6f7dbe9d" for claim "default/vut" saved
I1030 13:09:46.958799       1 controller.go:859] volume "pvc-9944178c-bd73-11e7-81db-021c6f7dbe9d" provisioned for claim "default/vut"
I1030 13:09:48.716027       1 leaderelection.go:198] stopped trying to renew lease to provision for pvc default/vut, task succeeded
^C
ubuntu@kubemaster-01:~/openebs/k8s$ kubectl delete -f test-pvc.yaml 
persistentvolumeclaim "vut" deleted
ubuntu@kubemaster-01:~/openebs/k8s$ k logs -f openebs-provisioner-3136189815-8wwnx 
I1030 13:09:11.887559       1 maya_api.go:65] Maya Cluster IP: 10.111.244.51
I1030 13:09:11.915624       1 controller.go:407] Starting provisioner controller 85a0b3b4-bd73-11e7-b51f-469266a6fddd!
I1030 13:09:46.677218       1 controller.go:1073] scheduleOperation[lock-provision-default/vut[9944178c-bd73-11e7-81db-021c6f7dbe9d]]
I1030 13:09:46.681262       1 controller.go:1073] scheduleOperation[lock-provision-default/vut[9944178c-bd73-11e7-81db-021c6f7dbe9d]]
I1030 13:09:46.692432       1 leaderelection.go:156] attempting to acquire leader lease...
I1030 13:09:46.703453       1 leaderelection.go:178] successfully acquired lease to provision for pvc default/vut
I1030 13:09:46.703677       1 controller.go:1073] scheduleOperation[provision-default/vut[9944178c-bd73-11e7-81db-021c6f7dbe9d]]
I1030 13:09:46.712277       1 maya_api.go:87] Volume Spec Created:
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: pvc-9944178c-bd73-11e7-81db-021c6f7dbe9d
  labels:
    volumeprovisioner.mapi.openebs.io/storage-size: 5G
    k8s.io/storage-class: openebs-standard
    k8s.io/namespace: default

I1030 13:09:46.833081       1 maya_api.go:115] Volume Successfully Created:
{"metadata":{"creationTimestamp":null,"name":"pvc-9944178c-bd73-11e7-81db-021c6f7dbe9d"},"spec":{"AccessModes":null,"Capacity":null,"ClaimRef":null,"OpenEBS":{"volumeID":""},"PersistentVolumeReclaimPolicy":"","StorageClassName":""},"status":{"Message":"","Phase":"","Reason":""}}
I1030 13:09:46.833168       1 maya_api.go:130] Get details for Volume :pvc-9944178c-bd73-11e7-81db-021c6f7dbe9d
I1030 13:09:46.948478       1 maya_api.go:148] Volume Details Successfully Retrieved
I1030 13:09:46.948632       1 controller.go:806] volume "pvc-9944178c-bd73-11e7-81db-021c6f7dbe9d" for claim "default/vut" created
I1030 13:09:46.958759       1 controller.go:823] volume "pvc-9944178c-bd73-11e7-81db-021c6f7dbe9d" for claim "default/vut" saved
I1030 13:09:46.958799       1 controller.go:859] volume "pvc-9944178c-bd73-11e7-81db-021c6f7dbe9d" provisioned for claim "default/vut"
I1030 13:09:48.716027       1 leaderelection.go:198] stopped trying to renew lease to provision for pvc default/vut, task succeeded
I1030 13:10:43.499463       1 controller.go:1073] scheduleOperation[delete-pvc-9944178c-bd73-11e7-81db-021c6f7dbe9d[996f0896-bd73-11e7-81db-021c6f7dbe9d]]
I1030 13:10:43.502203       1 maya_api.go:163] Delete Volume :pvc-9944178c-bd73-11e7-81db-021c6f7dbe9d
I1030 13:10:43.657337       1 maya_api.go:181] Volume Deleted Successfully initiated
I1030 13:10:43.657361       1 controller.go:1049] volume "pvc-9944178c-bd73-11e7-81db-021c6f7dbe9d" deleted
I1030 13:10:43.687424       1 controller.go:1060] volume "pvc-9944178c-bd73-11e7-81db-021c6f7dbe9d" deleted from database
^C

```

AmitKumarDas commented on openebs/openebs-provisioner.go:
```diff
@@ -51,6 +51,7 @@ type openEBSProvisioner struct {
+
 	nodeName := os.Getenv("NODE_NAME")
 	if nodeName == "" {
 		glog.Errorf("env variable NODE_NAME must be set so that this provisioner can identify itself")
```
The error message need not be a detailed description. It can be: ENV variable 'NODE_NAME' is not set
Check if there are any best practices for logging if you can.

AmitKumarDas commented on openebs/pkg/v1/maya_api.go:
```diff
@@ -82,14 +80,11 @@ func (v OpenEBSVolume) CreateVsm(vname, size, storageClassName string) (string,
 	yamlValue, _ := yaml.Marshal(vs)
 
-	glog.Infof("VSM Spec Created:\n%v\n", string(yamlValue))
+	glog.Infof("Volume Spec Created:\n%v\n", string(yamlValue))
```
This is a debug or may be more verbose level.

AmitKumarDas commented on openebs/pkg/v1/maya_api.go:
```diff
@@ -117,12 +112,12 @@ func (v OpenEBSVolume) CreateVsm(vname, size, storageClassName string) (string,
 
-	glog.Infof("VSM Successfully Created:\n%v\n", string(data))
-	return "VSM Successfully Created", nil
+	glog.Infof("Volume Successfully Created:\n%v\n", string(data))
```
If you are pushing the entire data then it should be a DEBUG. If just the name of volume, then INFO

AmitKumarDas commented on openebs/pkg/v1/maya_api.go:
```diff
@@ -132,7 +127,7 @@ func (v OpenEBSVolume) ListVsm(vname string, obj interface{}) error {
 	url := addr + "/latest/volumes/info/" + vname
 
-	glog.Infof("Get details for VSM :%v", string(vname))
+	glog.Infof("Get details for Volume :%v", string(vname))
```
This should be DEBUG or TRACE

AmitKumarDas commented on openebs/pkg/v1/maya_api.go:
```diff
@@ -150,12 +145,12 @@ func (v OpenEBSVolume) ListVsm(vname string, obj interface{}) error {
 		return err
 	}
-	glog.Info("VSM Details Successfully Retrieved")
+	glog.Info("Volume Details Successfully Retrieved")
```
Can the log statements follow certain standards ? Here everything is Capital cased.
Mention which volume details were retrieved successfully.

AmitKumarDas commented on openebs/pkg/v1/maya_api.go:
```diff
@@ -165,7 +160,7 @@ func (v OpenEBSVolume) DeleteVsm(vname string) error {
 	url := addr + "/latest/volumes/delete/" + vname
 
-	glog.Infof("Delete VSM :%v", string(vname))
+	glog.Infof("Delete Volume :%v", string(vname))
```
This should be TRACE

AmitKumarDas commented on openebs/pkg/v1/maya_api.go:
```diff
@@ -183,6 +178,6 @@ func (v OpenEBSVolume) DeleteVsm(vname string) error {
 		return err
 	}
-	glog.Info("VSM Deleted Successfully initiated")
+	glog.Info("Volume Deleted Successfully initiated")
```
Do we need to have every word as Capital cased ?
Mention the volume name which was targeted for deletion.

AmitKumarDas commented on openebs/types/v1/types.go:
```diff
@@ -16,15 +16,16 @@ limitations under the License.
 	Metadata   struct {
 		Name   string `yaml:"name"`
 		Labels struct {
 			Storage      string `yaml:"volumeprovisioner.mapi.openebs.io/storage-size"`
```
Can you raise an issue to send the Capacity as "openebs.io/capacity"
The storage-size should be deprecated in future & should remain for backward compatibility purposes.

Commit:
Fix error message in openebs-provisioner

This will fix error message in openebs-provisioner &
will send capacity as "openebs.io/capacity"

Commit:
Retain storage parameter in openebs-provisioner
